File: frontDemo/src/sim/precomputed.ts
/**
 * Precomputed upload results: the segmenter's output for a known file, made
 * earlier, used on request instead of a live run.
 *
 * WHY. The live segmenter takes seconds on WebGPU and tens of seconds on WASM,
 * and a demonstration should not stall on it. So the upload panel always offers
 * "Use precomputed result" (FUTURE_WORK.md §1.5): instant, and labelled as
 * precomputed wherever the result is shown.
 *
 * WHAT MAKES IT HONEST. A stored result is only ever the live result, stored:
 *
 *  - it is made OFFLINE by the same code the browser runs (`segment` in
 *    `segmenter.ts`, onnxruntime-web on WASM), by
 *    `scripts/precompute-uploads.ts`, and `check:precomputed` recomputes every
 *    entry whose file is on disk and demands the identical mask;
 *  - it is keyed by the SHA-256 of the file's bytes, so a renamed copy still
 *    finds it and a different file with the same name never does;
 *  - it records the model's SHA-256 and is refused for any other model, so a
 *    retrained release can never be shown a previous model's answer;
 *  - only the segmentation is stored. Decoding, tracing, the despeckled copy,
 *    the coastline and the drift all run live on the file in hand.
 *
 * The same contract is what the live pipeline (§3, `POST /runs`) has to honour:
 * one button, the precomputed answer for this exact input, and a label.
 */

import type { Detection, Segmentation } from "./segmenter";

export const PRECOMPUTED_DIR = "precomputed/";

export interface PrecomputedEntry {
  version: 1;
  /** The file name it was computed from, for the label. */
  file: string;
  /** Repository-relative path of that file, for provenance and the check. */
  source: string;
  /** SHA-256 of the file's bytes, hex. */
  sha256: string;
  bytes: number;
  width: number;
  height: number;
  model: { name: string; sha256: string };
  /** What ran it, stated on the result. */
  engine: string;
  computedAt: string;
  inferMs: number;
  tiles: number;
  detections: Detection[];
  /** The scene mask, row-major, as run lengths alternating 0 and 1, starting with 0. */
  maskRuns: number[];
  /** Anything about how the file was read that the label should carry. */
  decodeNote?: string;
}

export interface PrecomputedIndex {
  version: 1;
  /** File SHA-256 to the entry's file name in `PRECOMPUTED_DIR`. */
  entries: Record<string, string>;
}

export type Lookup = { ok: true; entry: PrecomputedEntry } | { ok: false; reason: string };

/** Run lengths of a 0/1 mask, alternating, starting with a run of zeros (possibly empty). */
export function encodeRuns(mask: Uint8Array): number[] {
  const runs: number[] = [];
  let value = 0;
  let length = 0;
  for (let i = 0; i < mask.length; i++) {
    const v = mask[i] ? 1 : 0;
    if (v === value) {
      length++;
      continue;
    }
    runs.push(length);
    value = v;
    length = 1;
  }
  runs.push(length);
  return runs;
}

export function decodeRuns(runs: number[], length: number): Uint8Array {
  const mask = new Uint8Array(length);
  let at = 0;
  for (let r = 0; r < runs.length; r++) {
    const n = runs[r];
    if (!Number.isInteger(n) || n < 0 || at + n > length) throw new Error("mask runs overrun the mask");
    if (r % 2 === 1) mask.fill(1, at, at + n);
    at += n;
  }
  if (at !== length) throw new Error(`mask runs cover ${at} of ${length} pixels`);
  return mask;
}

/** SHA-256 of some bytes, lower-case hex. WebCrypto: the browser and Node alike. */
export async function sha256Hex(data: ArrayBuffer | Uint8Array): Promise<string> {
  // A copy, so the digest never sees a view onto a larger shared buffer.
  const bytes = data instanceof Uint8Array ? data.slice() : new Uint8Array(data);
  const digest = await globalThis.crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest), (b) => b.toString(16).padStart(2, "0")).join("");
}

/** Whether a stored entry may stand in for this file under this model, and if not, why not. */
export function acceptEntry(
  entry: PrecomputedEntry,
  want: { sha256: string; modelSha256: string; width?: number; height?: number },
): Lookup {
  if (entry.version !== 1) return { ok: false, reason: `unknown precomputed format (version ${entry.version})` };
  if (entry.sha256 !== want.sha256) return { ok: false, reason: "the stored result is for a different file" };
  if (entry.model.sha256 !== want.modelSha256)
    return { ok: false, reason: "the stored result was made by a different model than the one loaded now" };
  if (want.width !== undefined && (entry.width !== want.width || entry.height !== want.height))
    return { ok: false, reason: `the stored result is ${entry.width} x ${entry.height}, the raster is ${want.width} x ${want.height}` };
  const total = entry.maskRuns.reduce((s, n) => s + n, 0);
  if (total !== entry.width * entry.height)
    return { ok: false, reason: `the stored mask covers ${total} of ${entry.width * entry.height} pixels` };
  return { ok: true, entry };
}

/** The stored result in the shape the rest of the upload path consumes. */
export function toSegmentation(entry: PrecomputedEntry): Segmentation {
  return {
    mask: decodeRuns(entry.maskRuns, entry.width * entry.height),
    width: entry.width,
    height: entry.height,
    detections: entry.detections,
    tiles: entry.tiles,
    ms: entry.inferMs,
  };
}

type JsonLoader = (path: string) => Promise<unknown>;

let loader: JsonLoader = async (path) => {
  const response = await fetch(path);
  // A static host answers a missing file with its index page; only JSON counts.
  if (!response.ok || !(response.headers.get("content-type") ?? "").includes("json")) return null;
  return response.json();
};

/** Swap the fetcher, for Node checks that read `public/` from disk. */
export function setPrecomputedLoader(next: JsonLoader) {
  loader = next;
}

/** The stored result for this file under this model, or why there is none. */
export async function findPrecomputed(sha256: string, modelSha256: string): Promise<Lookup> {
  const index = (await loader(`${PRECOMPUTED_DIR}index.json`)) as PrecomputedIndex | null;
  const name = index?.entries?.[sha256];
  if (!name) return { ok: false, reason: "no precomputed result exists for this file; it has to run live" };
  const entry = (await loader(`${PRECOMPUTED_DIR}${name}`)) as PrecomputedEntry | null;
  if (!entry) return { ok: false, reason: `the precomputed result ${name} could not be read` };
  return acceptEntry(entry, { sha256, modelSha256 });
}

File: backend/pipeline/precomputed.py
"""The §1.5 contract on the server: a stored segmentation for this exact input and model.

The console's "Use precomputed result" button (`frontDemo/src/sim/precomputed.ts`)
swaps a live segmenter run for one made earlier, and `POST /api/v1/runs` honours
the same contract with the same files:

  * keyed by the **SHA-256 of the file's bytes**, so a renamed copy still finds
    its entry and a different file with the same name never does;
  * stamped with the **model's SHA-256** and refused for any other model -- here
    both links of the chain are checked: the entry was made by the browser's
    ONNX export, and that export was made from the release weights the live
    backend path would run (`source_weights_sha256`);
  * **only the segmentation is swapped.** Seed choice, characterisation, CFAR,
    wind, drift and everything after run live on the file in hand, and every
    place the result is shown says PRECOMPUTED, by what, and when.

Entries are the browser segmenter's WASM output (`npm run precompute:uploads`),
which `check:precomputed` holds to the live browser run pixel for pixel.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import numpy as np

from backend.config import REPO_ROOT

PRECOMPUTED_DIR = REPO_ROOT / "frontDemo" / "public" / "precomputed"
ONNX_MANIFEST = REPO_ROOT / "frontDemo" / "public" / "models" / "L1-ciou-research.json"


@dataclass(frozen=True, slots=True)
class Lookup:
    entry: dict[str, Any] | None
    reason: str

    @property
    def ok(self) -> bool:
        return self.entry is not None


def find_precomputed(
    sha256: str,
    *,
    weights_sha256: str,
    directory: Path = PRECOMPUTED_DIR,
    onnx_manifest: Path = ONNX_MANIFEST,
) -> Lookup:
    """The stored segmentation for these bytes under the release model, or why there is none."""
    index_path = directory / "index.json"
    if not index_path.exists():
        return Lookup(None, "no precomputed results exist on this machine")
    name = json.loads(index_path.read_text(encoding="utf-8")).get("entries", {}).get(sha256)
    if not name:
        return Lookup(None, "no precomputed result exists for this file; it has to run live")
    entry = json.loads((directory / name).read_text(encoding="utf-8"))
    if entry.get("version") != 1:
        return Lookup(None, f"unknown precomputed format (version {entry.get('version')})")
    if entry.get("sha256") != sha256:
        return Lookup(None, "the stored result is for a different file")
    if not onnx_manifest.exists():
        return Lookup(None, "the browser model's manifest is missing, so the stored result's model cannot be checked")
    onnx = json.loads(onnx_manifest.read_text(encoding="utf-8"))
    if entry["model"]["sha256"] != onnx.get("sha256"):
        return Lookup(None, "the stored result was made by a different model than the browser's current one")
    if onnx.get("source_weights_sha256") != weights_sha256:
        return Lookup(None, "the browser model was not exported from the release weights this backend runs")
    total = int(sum(entry["maskRuns"]))
    if total != entry["width"] * entry["height"]:
        return Lookup(None, f"the stored mask covers {total} of {entry['width'] * entry['height']} pixels")
    return Lookup(entry, "")


def decode_runs(runs: list[int], length: int) -> np.ndarray:
    """Run lengths alternating 0 and 1, starting with 0 (`precomputed.ts` `decodeRuns`)."""
    mask = np.zeros(length, dtype=np.uint8)
    at = 0
    for r, n in enumerate(runs):
        if n < 0 or at + n > length:
            raise ValueError("mask runs overrun the mask")
        if r % 2 == 1:
            mask[at:at + n] = 1
        at += n
    if at != length:
        raise ValueError(f"mask runs cover {at} of {length} pixels")
    return mask


def entry_detections(entry: dict[str, Any], transform: Any) -> dict[str, Any]:
    """The stored mask as a detection FeatureCollection in lon/lat, shaped as `infer_scene`'s.

    The mask is the union of every detection's mask, so each connected region
    becomes one feature carrying the highest score among the detections whose
    box it overlaps -- the one stored figure that belongs to it.
    """
    from shapely.affinity import affine_transform
    from shapely.geometry import box, mapping

    from backend.detect.yolo_lsk.infer import mask_geometry

    width, height = int(entry["width"]), int(entry["height"])
    mask = decode_runs(entry["maskRuns"], width * height).reshape(height, width).astype(bool)
    geometry = mask_geometry(mask)
    parts = [] if geometry.is_empty else list(getattr(geometry, "geoms", [geometry]))
    boxes = [(box(*d["box"]), float(d["score"])) for d in entry.get("detections", [])]
    coefficients = [transform.a, transform.b, transform.d, transform.e, transform.c, transform.f]
    features = []
    for index, part in enumerate(parts):
        scores = [score for b, score in boxes if b.intersects(part)]
        features.append({
            "type": "Feature",
            "id": index,
            "geometry": mapping(affine_transform(part, coefficients)),
            "properties": {
                "class_name": "slick",
                "confidence": max(scores) if scores else 0.0,
                "research_only": True,
            },
        })
    return {"type": "FeatureCollection", "features": features}
